### ISSUE TITLE
bash scripts are treated as Python for some reason

### ISSUE THREAD

**Issue by wimglenn** [pypa/setuptools#1178]
When using the [`scripts` keyword argument](http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-scripts-keyword-argument) to include non-python scripts, `pip install --editable .` still attempts to treats them as Python code even when they are clearly shell scripts.  

I've included a reproducible example here:
https://github.com/wimglenn/pynapple/tree/pypa_issue

Script 1 is:
```
#!/usr/bin/bash

ls
```

Script 2 is:
```
#!/usr/bin/bash
:
ls
```

The only difference being I added a do-nothing `:` into the script which makes the file invalid syntax for Python.  

```
$ pip install --quiet --editable .
$ script2.sh   # works
LICENSE  pynapple  pynapple.egg-info  setup.cfg  setup.py  shell_scripts  venv
$ script1.sh   # crash
Traceback (most recent call last):
  File "/home/wglenn/git/pynapple/venv/bin/script1.sh", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/wglenn/git/pynapple/shell_scripts/script1.sh", line 3, in <module>
    ls
NameError: name 'ls' is not defined
```

The "installed" script2.sh is correct.  But the installed script1 has been wrapped in a python exec/compile!

```
$ cat ./venv/bin/script1.sh
#!/home/wglenn/git/pynapple/venv/bin/python3.6
# EASY-INSTALL-DEV-SCRIPT: 'pynapple==0.1','script1.sh'
__requires__ = 'pynapple==0.1'
__import__('pkg_resources').require('pynapple==0.1')
__file__ = '/home/wglenn/git/pynapple/shell_scripts/script1.sh'
exec(compile(open(__file__).read(), __file__, 'exec'))

```

I would have thought the shebang should be enough, but is there a better way to prevent EASY-INSTALL-DEV-SCRIPT from trying to wrap my shell scripts?

**Comment by htgoebel** [comment]
This can be closed as invalid: You are using
```
    scripts=[
        'shell_scripts/script1.sh',
        'shell_scripts/script2.sh',
    ],
```
while this [is for Pythonscripts](https://docs.python.org/3.9/distutils/setupscript.html#installing-scripts) only: "Scripts are files containing Python source code …"

**Comment by wimglenn** [comment]
Hmm, that is interesting!  Thought it was fairly common to use `scripts` for scripts such as shell scripts. Because for Python code, we just use `console_scripts` entry point group instead, and setuptools will auto-generate a script wrapper. So, distutils aside, isn't non-Python code the most convincing use-case for using `scripts` in setuptools?

**Comment by wimglenn** [comment]
I can confirm that using `pyproject.toml` and the `script-files` config of setuptools' config section works as expected:

```
[tool.setuptools]
packages = ["pynapple"]
script-files = [
    "shell_scripts/script1.sh",
    "shell_scripts/script2.sh",
]
```

I tested this with regular installs, built wheel, using both `pip` and `uv pip` installers, and did not find any problems. So whatever caused the `EASY-INSTALL-DEV-SCRIPT` issue wrapping shell scripts as Python, was either a problem specific to `setup.py` code-paths, or has been since fixed.

I don't know why the setuptools docs [mark script-files as **"Discouraged"**](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration), because like you said it's not possible to use `project.scripts` here  and there isn't a working alternative afaik. I am still using this feature for shell script packaging, so I sure hope they don't deprecate/remove it without an acceptable replacement.

If anyone knows about possible future developments here it would probably be @abravalheri .

**Comment by wimglenn** [comment]
Would you be open to softening the language in the docs around **script-files** (and possibly **data-files**, too), making it clear there's not any danger of being deprecated? At least to me it is not obvious that the note intends mainly to discourage against using these for Python scripts - I would always reach for entrypoints instead of scripts for Python anyway.

I think this issue could be closed in that case. The problem is no longer reproducible with recent versions of pip/uv and setuptools.

**Comment by abravalheri** [comment]
Isn't that a bit of overthinking?

I can see where you're coming from and I appreciate your perspective.
But the term "discouraged" in the documentation isn't meant to imply that these features are on the verge of deprecation. On the contrary, it is quite literal: we are basically saying: "we encourage people to think about other alternatives before considering this feature".

It's intended to guide users, especially beginners, towards best practices. We've found that some features can be a bit tricky, and we want to help users to avoid potential pitfalls. The term "discouraged" is hint that there might be better, more efficient options available for a big part of the use cases.

It's worth noting that any feature in setuptools can potentially be deprecated in the future. Being marked as "discouraged" doesn't change that probability in any way. In fact, since I wrote the `pyproject.toml` documentation, there have been other features that were deprecated and removed without ever being documented as "discouraged".

The association between the term "discouraged" and "pre-deprecation period" does not seem justified. If it sounds intimidating, that is probably because it is proportional to the reports of users having trouble to use the feature.
